drivers/SmartThings/matter-switch/src/init.lua

```diff
@@ -121,98 +270,233 @@ local function assign_child_profile(device, child_ep)
-  -- do not have a generic fingerprint and will join as a matter-thing. However, we have seen some devices
-  -- claim to be Light Switch device types and still implement their clusters as server, so this is a
-  -- workaround for those devices.
-  if num_server_eps > 0 and detect_matter_thing(device) == true then
-    local id = 0
-    for _, ep in ipairs(device.endpoints) do
-      -- main_endpoint only supports server cluster by definition of get_endpoints()
-      if main_endpoint == ep.endpoint_id then
-        for _, dt in ipairs(ep.device_types) do
-          -- no device type that is not in the switch subset should be considered.
-          if (ON_OFF_SWITCH_ID <= dt.device_type_id and dt.device_type_id <= ON_OFF_COLOR_DIMMER_SWITCH_ID) then
-            id = math.max(id, dt.device_type_id)
+    if num_server_eps > 1  then
+      -- If the device is a parent child device, then set the find_child function on init.
+      -- This is persisted because initialize switch is only run once, but find_child function should be set
+      -- on each driver init.
+      device:set_field(IS_PARENT_CHILD_DEVICE, true, {persist = true})
+    end
+
+    device:set_field(SWITCH_INITIALIZED, true)
+
+    if #button_eps > 0 then
+      local battery_support = false
+      if device.manufacturer_info.vendor_id ~= HUE_MANUFACTURER_ID and
+          #device:get_endpoints(clusters.PowerSource.ID, {feature_bitmap = clusters.PowerSource.types.Feature.BATTERY}) > 0 then
+        battery_support = true
+      end
+      if tbl_contains(STATIC_PROFILE_SUPPORTED, #button_eps) then
+        if battery_support then
+          new_profile = string.format("%d-button-battery", #button_eps)
+        else
+          new_profile = string.format("%d-button", #button_eps)
+        end
+      elseif not battery_support then
+        -- a battery-less button/remote (either single or will use parent/child)
+        new_profile = "button"
+      end
+
+      if new_profile then
+        if #switch_eps > 0 then
+          new_profile = new_profile .. "-switch"
+        end
+        device:try_update_metadata({profile = new_profile})
+      end
+    else
+      -- The case where num_server_eps > 0 is a workaround for devices that have a
+      -- Light Switch device type but implement the On Off cluster as server (which is against the spec
+      -- for this device type). By default, we do not support Light Switch device types because by spec these
+      -- devices need bindings to work correctly (On/Off cluster is client in this case), so these device types
+      -- do not have a generic fingerprint and will join as a matter-thing. However, we have seen some devices
+      -- claim to be Light Switch device types and still implement their clusters as server, so this is a
+      -- workaround for those devices.
+      if num_server_eps > 0 and detect_matter_thing(device) == true then
```

Comment:
You'll what to make this `num_switch_server_eps` so that we are not counting button server eps also. This logic should only affect Light Switch device types as described in the comment above, so we only want to count server eps for the light switch types when determining if we need to go down this codepath.